protocol/kafka_confluent/v2/message.go

```diff
@@ -0,0 +1,146 @@
+/*
+ Copyright 2023 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package kafka_confluent
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/binding/format"
+	"github.com/cloudevents/sdk-go/v2/binding/spec"
+	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
+)
+
+const (
+	prefix         = "ce-"
+	contentTypeKey = "content-type"
+)
+
+const (
+	KafkaOffsetKey    = "kafkaoffset"
+	KafkaPartitionKey = "kafkapartition"
+	KafkaTopicKey     = "kafkatopic"
+	KafkaMessageKey   = "kafkamessagekey"
+)
+
+var specs = spec.WithPrefix(prefix)
+
+// Message represents a Kafka message.
+// This message *can* be read several times safely
+type Message struct {
+	internal   *kafka.Message
+	properties map[string][]byte
+	format     format.Format
+	version    spec.Version
+}
+
+// Check if Message implements binding.Message
+var (
+	_ binding.Message               = (*Message)(nil)
+	_ binding.MessageMetadataReader = (*Message)(nil)
+)
+
+func NewMessage(msg *kafka.Message) *Message {
+	var contentType, contentVersion string
+	properties := make(map[string][]byte, len(msg.Headers)+3)
+	for _, header := range msg.Headers {
+		k := strings.ToLower(string(header.Key))
+		if k == strings.ToLower(contentTypeKey) {
+			contentType = string(header.Value)
+		}
+		if k == specs.PrefixedSpecVersionName() {
+			contentVersion = string(header.Value)
+		}
+		properties[k] = header.Value
+	}
+
+	// add the kafka message key, topic, partition and partition key to the properties
+	properties[prefix+KafkaOffsetKey] = []byte(strconv.FormatInt(int64(msg.TopicPartition.Offset), 10))
+	properties[prefix+KafkaPartitionKey] = []byte(strconv.FormatInt(int64(msg.TopicPartition.Partition), 10))
+	if msg.TopicPartition.Topic == nil {
+		panic("the kafka message topic shouldn't be nil")
+	}
```

Comment:
note: the nil check should come at the beginning of the function and check whether:

- `msg` is nil
- `msg.TopicPartition.Topic` is nil

Reply:
Done.